### PR TITLE
Fixed bug from #152.

### PR DESCRIPTION
**Pull request by KamilKrol**
Problem with overriden Enum serialization to Map.
We added also tests confirming the problem.

Kamil Krol & Dreando

### CODE REVIEW

**Comment by garethahealy** [comment]
@KamilKrol ; can you rebase to force re-test please

**Comment by garethahealy** [comment]
@KamilKrol ; any update?

**Comment by KamilKrol** [comment]
@garethahealy , done. Let me know if you need something additional before merging this pull request.

Kamil

**Comment by garethahealy** [comment]
@orange-buffalo ; can you review please

**Comment by garethahealy** [comment]
@KamilKrol ; instead of using "expected=Exception", can you use the JUnit rule checker instead.
- https://github.com/DozerMapper/dozer/blob/master/core/src/test/java/org/dozer/functional_tests/GranularDozerBeanMapper_WithExceptionLoggedTest.java#L36

As currently, MappingException is for everything, so we wouldn't know if this test was actually broken in the future.

File: core/src/main/java/org/dozer/MappingProcessor.java [hunk 1]
```diff
@@ -517,7 +517,21 @@ private Object mapOrRecurseObject(Object srcObj, Object srcFieldValue, Class<?>
...
+      /* This cast is case because of the check performed by the checkIfOverridenEnum function. */
+      @SuppressWarnings("unchecked")
+      Class<T> castedSuperclass = (Class<T>) enumClass.getSuperclass();
+      return castedSuperclass;
+    }
+    return enumClass;
+  }
+
+  private <T extends Enum<T>> boolean checkIfOverridenEnum(Class<T> enumClass) {
+    return Enum.class.isAssignableFrom(enumClass.getSuperclass()) && !Enum.class.equals(enumClass.getSuperclass());
```
Comment (orange-buffalo) [inline core/src/main/java/org/dozer/MappingProcessor.java]: `Enum.class.isAssignableFrom(enumClass.getSuperclass())` is redundant. It is already guaranteed that this code will be executed only for enums (see`MappingUtils.isEnumType` before `mapEnum` call).

**Comment by garethahealy** [comment]
@KamilKrol ; are you able to work on the suggested changes?

**Comment by KamilKrol** [comment]
@garethahealy , I fixed all comments from code review and committed latest version.
But it seems that some OSGI tests are failing.

**Comment by garethahealy** [comment]
@KamilKrol ; the OSGi failure was a test flake.

**Comment by garethahealy** [comment]
LGTM

@orange-buffalo ?

**Comment by garethahealy** [comment]
https://github.com/DozerMapper/dozer/issues/152

**Comment by KamilKrol** [comment]
@garethahealy , perfect that this OSGI failure is understood.
Let me know if I can do anything more for this pull requests.

**Pull request merged**